src/accelerate/data_loader.py

```diff
@@ -580,6 +571,22 @@ def __iter__(self):
         self.iteration += 1
         self.end()
 
+    def __reduce__(self):
+        return (
+            DataLoaderShard,
+            (
+                self.base_dataloader.dataset,
+                self.device,
+                self.rng_types,
+                self.synchronized_generator,
+                self.skip_batches,
+                self.use_stateful_dataloader,
+                self._drop_last,
+                self._non_blocking,
+            ),
+            self.__dict__,
```

Comment:
Instead of hard-coding these arguments, how about calling `super().__reduce__()`?

Reply:
I don't know if that works. IIUC, `__reduce__` is called to specify how we want to pickle an object for reconstruction. In this context, if I'm understanding right, the current code states "We want to pickle a DataLoaderShard by constructing a DataLoaderShard with these arguments and filling it's `__dict__` with my elements."

If we replace this with `return super().__reduce__()`, then doesn't this change this to "We want to pickle a DataLoaderShard by turning it into a DataLoader." In that case, the following would happen:

```python
## Behavior of code using super().__reduce__():

# accelerator prepares a dataloader shard from a dataloader
dl_shard = original_accel.prepare(dl)
assert isinstance(dl_shard, DataLoaderShard)
assert isinstance(original_accel._dataloaders[0], DataLoaderShard)

# Try to pickle and unpickle the dataloader
loaded_accel = pickle.loads(pickle.dumps(original_accel)

# The unpickled accelerator's underlying dataloaders are no longer sharded
assert isinstance(loaded_accel._dataloaders[0], DataLoader)
```
Is this the intended behavior?

Reply:
Sorry, what I meant is something like:

```python
args = super().__reduce__()
return (DataLoaderShard, *args[1:])
```

That way, the 2nd and 3rd argument (which are roughly the arguments needed to construct the instance) don't need to be hardcoded, only the constructor class. If we hard-code the arguments, we need to remember to change them each time they're changed on the class.

Reply:
Done. The above block doesn't include the object `__dict__`, but I'm not sure if we need to add that or not. It passes my existing tests as written